### ISSUE TITLE
TypeScript definition

### ISSUE THREAD

**Issue by mattheworiordan** [ably/ably-js#320]
A few customers have requested a TypeScript definition.  We should look into providing this given [we have a very well documented API anyway](https://www.ably.io/documentation)

**Comment by mattheworiordan** [comment]
First attempt from a customer at defining a TypeScript definition below.  Clearly it's far from exhaustive, but gives us a starting point.

Client

```
declare namespace __Ably {
  interface Message<T> {
    name: string;
    data: T;
    id: string;
    timestamp: number;
    connectionId: number;
  }

  interface EventCallback<TMessage> {
    (message: Message<TMessage>): any;
  }

  interface Channel {
    subscribe: (eventName: string, eventCallback: EventCallback<any>) => any;
    publish: (eventName: string, data: any) => any;
  }

  interface Channels {
    get: (channelName: string) => Channel;
  }

  interface RealtimeClass {
    new({key: string}): Realtime;
  }

  interface Realtime {
    channels: Channels;
  }
}

interface Ably {
  Realtime: __Ably.RealtimeClass;
}

declare var Ably: Ably;
```

Server

```
declare module 'ably' {

  interface Channel {
    publish(event: string, data: Object, callback: (err) => void);
  }

  interface Channels {
    get(channelName: string): Channel;
  }

  interface Rest {
    new ({key: string}): Rest;
    channels: Channels;
  }

  interface Ably {
    Rest: Rest;
  }

  let Ably: Ably
  export = Ably;
}
````

**Comment by mattheworiordan** [comment]
@SimonWoolf do you think once your work on 0.9 is complete we should release a formal TypeScript definition file.  It would certainly help Giovanni's efforts as well, and I know a few customers are looking for this.  It shouldn't be too difficult given we do [have a very clear IDL anyway](http://docs.ably.io/client-lib-development-guide/features/#idl)

**Comment by wichert** [comment]
FWIW the typescript typings currently included in ably-js are incorrect: Ably.ablyLib.connection.once() is declared to take an connectionEventCallback type. That type specifies a callback taking an event and a state change parameter, while in reality the callback only receives a single Ably.ablyLib.ConnectionStateChange parameter.

**Comment by mattheworiordan** [comment]
@wichert thanks for reporting.  I've created a separate issue #443 to get this fixed.  We're mostly on holidays right now, but we'll try and get someone to look at this in the new year.

**Comment by wichert** [comment]
No worries - it's pretty easy to work around with a type cast in the meantime.

**Comment by SimonWoolf** [comment]
@wichert Would you mind trying out the typings in https://github.com/ably/ably-js/pull/444 and telling me if it fixes your issue? Thanks :)

**Comment by wichert** [comment]
Hi Simon,

I’m currently traveling. I intend to try the updates next week.
